File: app/core/ids/ulid.py
"""Crockford Base32 ULID validator (spec/21-app/20 §Identifier contract).

Length 26, alphabet [0-9A-HJKMNP-TV-Z]. No decoding — validation only.
Used by every boundary that accepts a jobId/taskId/ruleId route param
(F-20/F-21/F-29) so downstream code never sees a malformed identifier.
"""
from __future__ import annotations

import re

from app.core.errors.codes import ErrorCode

# Crockford Base32: excludes I, L, O, U to avoid transcription errors.
ULID_RE = re.compile(r"^[0-9A-HJKMNP-TV-Z]{26}$")


class UlidFormatError(ValueError):
    """Raised when a value is not a well-formed ULID."""

    code = ErrorCode.E_RULE_BAD_INPUT  # boundary reuse; caller may rewrap

    def __init__(self, value: object, field: str = "id") -> None:
        super().__init__(f"{field}={value!r} is not a valid ULID")
        self.field = field
        self.value = value


def is_ulid(value: object) -> bool:
    return isinstance(value, str) and ULID_RE.match(value) is not None


def assert_ulid(value: object, field: str = "id") -> str:
    if not is_ulid(value):
        raise UlidFormatError(value, field)
    return value  # type: ignore[return-value]

File: src/lib/ids/ulid.ts
/**
 * Crockford Base32 ULID validator - mirrors `app/core/ids/ulid.py`.
 * Every route param / RPC boundary that accepts a jobId/taskId/ruleId
 * runs the value through `assertUlid` so downstream renderers never see
 * malformed identifiers (audit F-20/F-21/F-29).
 */

export const ULID_RE = /^[0-9A-HJKMNP-TV-Z]{26}$/;

export function isUlid(value: unknown): value is string {
  return typeof value === "string" && ULID_RE.test(value);
}

export class UlidFormatError extends Error {
  readonly code = "E_ID_INVALID" as const;
  readonly field: string;
  constructor(value: unknown, field = "id") {
    super(`${field}=${JSON.stringify(value)} is not a valid ULID`);
    this.name = "UlidFormatError";
    this.field = field;
  }
}

export function assertUlid(value: unknown, field = "id"): string {
  if (isUlid(value) === false) throw new UlidFormatError(value, field);

  return value;
}
